flake8 reports F811 redefinition of unused 'step_impl'

Ignore F811 for duplicate `step_impl` functions ...
-------------------------

I'm using flake8 to check my source code, and I've been ignoring the typical [F811 redefinition violation](http://flake8.pycqa.org/en/latest/user/error-codes.html) globally in the past. I'd really prefer not to ignoring violations at all, so I've browsed the [behave docs](https://pythonhosted.org/behave/search.html?q=step_impl&check_keywords=yes&area=default) whether using the default `step_impl()` function name is actually the recommended approach.

... to silence false alarms
-------------------

While I've not seen an explicit confirmation of this I've seen `step_impl` being used exclusively for all occurrences of feature step implementations. I understand (or rather believe) that the behave decorators change the signature of the `step_impl` function, so that technically they are all different functions even if they have the same name in the source code (hence "redefinition" is a false alarm). I also believe that's the reason why Pylint doesn't report this as an error, because flake8 performs static checking only (hence has its limits).

Don't import decorators (... and get more violations)
------------------------------

Interestingly, reading the docs I found the following paragraph in the [API chapter](http://behave.readthedocs.io/en/latest/api.html#step-functions):

> _You don’t need to import the decorators_: they’re automatically available to your step implementation modules as global variables. (dead link)

Following this advice I obviously get another violation reported:

``` bash
$ flake8 
./tests/acceptance/steps/given.py:9:2: F821 undefined name 'given'
./tests/acceptance/steps/given.py:26:1: F811 redefinition of unused 'step_impl' from line 9
./tests/acceptance/steps/given.py:26:2: F821 undefined name 'given'
./tests/acceptance/steps/when.py:7:2: F821 undefined name 'when'
./tests/acceptance/steps/then.py:6:2: F821 undefined name 'then'
```
Maybe it's of interest to note that I had `from behave import given` (etc.) in the implementation modules; I prefer not to do the discouraged `from behave import *` import.

What's the right way?
-----------

Now, if I wanted to do it all right what should I do (to make the code PEP8 compliant, obviously)?

1. Give every `step_impl` function a unique name, manually? (... and update the behave docs)
1. ... and do behave imports? (... and update the behave docs)
1. Or write a flake8 plugin or configuration recommendation to silence the violations on valid step implementation code?

--- a/docs/tutorial.rst
+++ b/docs/tutorial.rst
@@ -278,6 +278,14 @@ or "then". The "and" and "but" step types are renamed internally to take
 the preceding step's keyword (so an "and" following a "given" will become a
 "given" internally and use a **given** decorated step).
 
+.. note::
+
+  Step function names do not need to have a unique symbol name, because the
+  text matching selects the step function from the step registry before it is
+  called as anonymous function.  Hence, when *behave* prints out the missing
+  step implementations in a test run, it uses "step_impl" for all functions
+  by default.
+
 If you find you'd like your step implementation to invoke another step you
 may do so with the :class:`~behave.runner.Context` method
 :func:`~behave.runner.Context.execute_steps`.

--- a/docs/tutorial.rst
+++ b/docs/tutorial.rst
@@ -23,19 +23,19 @@ file called "tutorial.py" containing:
 
 .. code-block:: python
 
-  from behave import *
+    from behave import *
 
-  @given('we have behave installed')
-  def step_impl(context):
-      pass
+    @given('we have behave installed')
+    def step_impl(context):
+        pass
 
-  @when('we implement a test')
-  def step_impl(context):
-      assert True is not False
+    @when('we implement a test')
+    def step_impl(context):
+        assert True is not False
 
-  @then('behave will test it for us!')
-  def step_impl(context):
-      assert context.failed is False
+    @then('behave will test it for us!')
+    def step_impl(context):
+        assert context.failed is False
 
 Run behave::
 
@@ -219,10 +219,10 @@ function. The table for the example above could be accessed like so:
 
 .. code-block:: python
 
-  @given('a set of specific users')
-  def step_impl(context):
-      for row in context.table:
-          model.add_user(name=row['name'], department=row['department'])
+    @given('a set of specific users')
+    def step_impl(context):
+        for row in context.table:
+            model.add_user(name=row['name'], department=row['department'])
 
 There's a variety of ways to access the table data - see the
 :class:`~behave.model.Table` API documentation for the full details.
@@ -258,20 +258,20 @@ Step code implementing the two steps here might look like
 
 .. code-block:: python
 
- @given('I search for a valid account')
- def step_impl(context):
-    context.browser.get('http://localhost:8000/index')
-    form = get_element(context.browser, tag='form')
-    get_element(form, name="msisdn").send_keys('61415551234')
-    form.submit()
+    @given('I search for a valid account')
+    def step_impl(context):
+        context.browser.get('http://localhost:8000/index')
+        form = get_element(context.browser, tag='form')
+        get_element(form, name="msisdn").send_keys('61415551234')
+        form.submit()
 
- @then('I will see the account details')
- def step_impl(context):
-    elements = find_elements(context.browser, id='no-account')
-    eq_(elements, [], 'account not found')
-    h = get_element(context.browser, id='account-head')
-    ok_(h.text.startswith("Account 61415551234"),
-        'Heading %r has wrong text' % h.text)
+    @then('I will see the account details')
+    def step_impl(context):
+        elements = find_elements(context.browser, id='no-account')
+        eq_(elements, [], 'account not found')
+        h = get_element(context.browser, id='account-head')
+        ok_(h.text.startswith("Account 61415551234"),
+            'Heading %r has wrong text' % h.text)
 
 The ``step`` decorator matches the step to *any* step type, "given", "when"
 or "then". The "and" and "but" step types are renamed internally to take
@@ -319,10 +319,10 @@ clauses (with a Given step that puts some text into
 
 .. code-block:: python
 
- @then('the result page will include "{text}"')
- def step_impl(context, text):
-    if text not in context.response:
-        fail('%r not in %r' % (text, context.response))
+    @then('the result page will include "{text}"')
+    def step_impl(context, text):
+        if text not in context.response:
+            fail('%r not in %r' % (text, context.response))
 
 There are several parsers available in *behave* (by default):
 
@@ -391,15 +391,15 @@ steps you define you might have:
 
 .. code-block:: python
 
-  @given('I request a new widget for an account via SOAP')
-  def step_impl(context):
-      client = Client("http://127.0.0.1:8000/soap/")
-      context.response = client.Allocate(customer_first='Firstname',
-          customer_last='Lastname', colour='red')
+    @given('I request a new widget for an account via SOAP')
+    def step_impl(context):
+        client = Client("http://127.0.0.1:8000/soap/")
+        context.response = client.Allocate(customer_first='Firstname',
+            customer_last='Lastname', colour='red')
 
-  @then('I should receive an OK SOAP response')
-  def step_impl(context):
-      eq_(context.response['ok'], 1)
+    @then('I should receive an OK SOAP response')
+    def step_impl(context):
+        eq_(context.response['ok'], 1)
 
 There's also some values added to the context by *behave* itself:
 
@@ -457,26 +457,26 @@ server and browser to run all your tests in. For example:
 
 .. code-block:: python
 
-  import threading
-  from wsgiref import simple_server
-  from selenium import webdriver
-  from my_application import model
-  from my_application import web_app
+    import threading
+    from wsgiref import simple_server
+    from selenium import webdriver
+    from my_application import model
+    from my_application import web_app
 
-  def before_all(context):
-      context.server = simple_server.WSGIServer(('', 8000))
-      context.server.set_app(web_app.main(environment='test'))
-      context.thread = threading.Thread(target=context.server.serve_forever)
-      context.thread.start()
-      context.browser = webdriver.Chrome()
+    def before_all(context):
+        context.server = simple_server.WSGIServer(('', 8000))
+        context.server.set_app(web_app.main(environment='test'))
+        context.thread = threading.Thread(target=context.server.serve_forever)
+        context.thread.start()
+        context.browser = webdriver.Chrome()
 
-  def after_all(context):
-      context.server.shutdown()
-      context.thread.join()
-      context.browser.quit()
+    def after_all(context):
+        context.server.shutdown()
+        context.thread.join()
+        context.browser.quit()
 
-  def before_feature(context, feature):
-      model.init(environment='test')
+    def before_feature(context, feature):
+        model.init(environment='test')
 
 Of course if you wish you could have a new browser for each feature, or to
 retain the database state between features or even initialise the database
@@ -549,20 +549,20 @@ browser and web server then you could tag them ``@browser``:
 
 .. code-block:: python
 
-  def before_feature(context, feature):
-      model.init(environment='test')
-      if 'browser' in feature.tags:
-          context.server = simple_server.WSGIServer(('', 8000))
-          context.server.set_app(web_app.main(environment='test'))
-          context.thread = threading.Thread(target=context.server.serve_forever)
-          context.thread.start()
-          context.browser = webdriver.Chrome()
+    def before_feature(context, feature):
+        model.init(environment='test')
+        if 'browser' in feature.tags:
+            context.server = simple_server.WSGIServer(('', 8000))
+            context.server.set_app(web_app.main(environment='test'))
+            context.thread = threading.Thread(target=context.server.serve_forever)
+            context.thread.start()
+            context.browser = webdriver.Chrome()
 
-  def after_feature(context, feature):
-      if 'browser' in feature.tags:
-          context.server.shutdown()
-          context.thread.join()
-          context.browser.quit()
+    def after_feature(context, feature):
+        if 'browser' in feature.tags:
+            context.server.shutdown()
+            context.thread.join()
+            context.browser.quit()
 
 
 Works In Progress
@@ -632,5 +632,3 @@ by using the user-specific configuration data. A user can:
             # NOTE: Use IPython debugger, same for pdb (basic python debugger).
             import ipdb
             ipdb.post_mortem(step.exc_traceback)
-
-

